### PR TITLE
Update sequence impl to tune controller memory consumption

### LINKED ISSUE #2386: the sequence impl drags memory
for the duration of a sequence activation, the controller "drags" the memory of the input payload, and *all* of the intermediate responses.

if done properly, the sequence controller should only need to keep around the response of the prior component of the sequence, i.e. the max memory consumption should be the maximum payload size of any component (or the input to the sequence); whereas right now, the max memory consumption of a sequence is the sum of all payloads, including the input to the sequence.

### PR DESCRIPTION
**Pull request by starpit**
Fixes #2386

This PR changes SequenceActions to use foldLeft rather than scanLeft. The change requires the introduction of a new accumulator data structure, to maintain the accumulated state for the reduction. The accumulated state includes an array of handles to logs (for the components), the cumulative duration, and a few other bits needed for creation of the final sequence activation record.

This fix also patches a closely related memory drag, caused by the way timeouts of sequence invocations are handled. See #2294 which patches other parts of the controller for the same kind of timeout drag. This patch needs to be a little different, due to the complex way the SequenceActions impl uses nested Promises.

PG3 625 🔵

### CODE REVIEW

**Comment by rabbah** [comment]
@tardieu @markusthoemmes @ioana-blue might any of you want to review this?

**Comment by rabbah** [comment]
@starpit as I noted on slack, we should have unit tests for the new/refactored code. It's far more testable now and we can get good coverage. If you don't have bandwidth for this, I will work on that tomorrow.

File: core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala [hunk 1]
```diff
@@ -365,55 +289,195 @@ protected[actions] trait SequenceActions {
...
+ */
+protected[actions] object SequenceAccounting {
+
+    def maxMemory(prevMemoryLimit: Option[Int], newMemoryLimit: Option[Int]): Option[Int] = {
+        prevMemoryLimit.map { prevMax =>
+            newMemoryLimit
+                .map(currentMax => Some(Math.max(prevMax, currentMax)))
+                .getOrElse(prevMemoryLimit)
+        }.getOrElse(newMemoryLimit)
+    }
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala]: Is this the same as:

```
(prevMemoryLimit ++ newMemoryLimit).reduceOption(Math.max)
```

File: core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala [hunk 2]
```diff
@@ -365,55 +289,195 @@ protected[actions] trait SequenceActions {
...
+ * @param duration the "user" time so far executing the sequence (sum of durations for
+ *        all actions invoked so far which is different from the total time spent executing the sequence)
+ * @param maxMemory the maximum memory annotation observed so far for the
+ *        components (needed to annotate the sequence with GB-s)
+ * @param shortcircuit when true, stops the execution of the next component in the sequence
+ */
+protected[actions] case class SequenceAccounting(
+    atomicActionCnt: Int,
+    previousResponse: AtomicReference[ActivationResponse],
+    logs: mutable.ArrayBuffer[ActivationId],
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala]: Do we rely on this being an **Array**Buffer anywhere, or could it aswell be the more general `Buffer`?

File: core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala [hunk 3]
```diff
@@ -365,55 +289,195 @@ protected[actions] trait SequenceActions {
...
+                    limitsAnnotation.asJsObject.getFields("memory") match {
+                        case Seq(JsNumber(memory)) => Some(memory.toInt) // we have a numerical "memory" field in the "limits" annotation
+                    }
+            } getOrElse { None })
+    }
+
+    /** The previous activation failed (this is used when there is no activation record or an internal error. */
+    def fail(failureResponse: ActivationResponse, activationId: Option[ActivationId]) = {
+        require(!failureResponse.isSuccess)
+        activationId.foreach(logs += _)
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/actions/SequenceActions.scala]: `logs.appendAll(activationId)`?

**Pull request merged**